util/rate_limiter.cc

```diff
@@ -136,96 +136,85 @@ void GenericRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
+  // - (1) is typically done by a thread at the front of a queue. This is
+  //   trivial when the queues are initially empty as the first choice that
+  //   arrives must be the only entry in its queue. When queues are initially
+  //   non-empty, we achieve this by having (2) notify a thread at the front
+  //   of a queue (preferring higher priority) to perform the next duty.
+  // - We do not require any additional wakeup for (2). Typically it will just
+  //   be done by the thread that finished (1).
+  //
+  // Combined, the second and third bullet points above suggest the refill/
+  // granting will typically be done by a request at the front of its queue.
+  // This is important because one wakeup is saved when a granted request
+  // happens to be in an already running thread.
+  //
+  // Note this nice property is not guaranteed in a few cases, however.
+  //
+  // - No request may be granted.
+  // - Requests from a different queue may be granted.
+  // - (2) may be run by a non-front request thread causing it to not be granted
+  //   even if some requests in that same queue are granted. It can happen for a
+  //   couple (unlikely) reasons.
+  //    - A new request may sneak in and grab the lock at the refill time,
+  //      before the thread finishing (1) can wake up and grab it.
+  //    - A new request may sneak in and grab the lock and execute (1) before
+  //      (2)'s chosen candidate can wake up and grab the lock. Then that non-
+  //      front request thread performing (1) can carry over to perform (2).
   do {
-    bool timedout = false;
-    // Leader election, candidates can be:
-    // (1) a new incoming request,
-    // (2) a previous leader, whose quota has not been not assigned yet due
-    //     to lower priority
-    // (3) a previous waiter at the front of queue, who got notified by
-    //     previous leader
-    if (leader_ == nullptr &&
-        ((!queue_[Env::IO_HIGH].empty() &&
-            &r == queue_[Env::IO_HIGH].front()) ||
-         (!queue_[Env::IO_LOW].empty() &&
-            &r == queue_[Env::IO_LOW].front()))) {
-      leader_ = &r;
-      int64_t delta = next_refill_us_ - NowMicrosMonotonic();
-      delta = delta > 0 ? delta : 0;
-      if (delta == 0) {
-        timedout = true;
+    int64_t time_until_refill_us = next_refill_us_ - NowMicrosMonotonic();
+    if (time_until_refill_us > 0) {
+      if (waiting_) {
+        // Somebody is performing (1). Trust we'll be woken up when our request
+        // is granted or we are needed for future duties.
+        r.cv.Wait();
       } else {
-        int64_t wait_until = clock_->NowMicros() + delta;
+        // Whichever thread reaches here first performs duty (1) as described
+        // above.
         RecordTick(stats, NUMBER_RATE_LIMITER_DRAINS);
         ++num_drains_;
-        timedout = r.cv.TimedWait(wait_until);
+        waiting_ = true;
+        r.cv.TimedWait(clock_->NowMicros() + time_until_refill_us);
```

Comment:
Maybe just me not seeing it - why the previous impl is a bug?

Reply:
Got you - yeah, I was looking at the commit history for the reason of this code change.  I personally prefer the option of revering a non-essential change like this, just to avoid potential future confusion when looking at this specific code change in this PR.

Question: I wonder - can we do "r.cv.TimedWait(next_refill_us_);" for a better semantics?

Reply:
> Question: I wonder - can we do "r.cv.TimedWait(next_refill_us_);" for a better semantics?

Not sure, `next_refill_us_` gets computed using `NowMicrosMonotonic()` while the existing code was specifying a time calculated from `NowMicros()`. I haven't learned what is the difference so stuck with what was there.

Reply:
Got you - then I am good with keeping your changes r.cv.TimedWait(clock_->NowMicros() + time_until_refill_us);.OR reverting this non-essential change. They don't seem much difference to me in correctness and reverting this non-essential is only for future reader of this PR.

Reply:
Reverted.